WNPRC_EHR/src/org/labkey/wnprc_ehr/notification/BloodAlertsNotificationRevamp.java

```diff
@@ -0,0 +1,482 @@
+                        if (!summaryArea.isEmpty()) {
+                            messageBody.append("<b>" + summaryArea + ":</b><br>\n");
+                        }
+                        else {
+                            messageBody.append("<b>" + "Unknown Area" + ":</b><br>\n");
+                        }
+                        //Sorts currentSummaryArea by room key.
+                        HashMap<String,bloodDrawRoom> currentSummaryArea = myBloodDrawAlertObject.summaryHashMap.get(summaryArea);
+                        //Iterates through each room.
+                        for (String summaryRoom : notificationToolkit.sortSetWithNulls(currentSummaryArea.keySet())) {
+                            bloodDrawRoom currentSummaryRoom = currentSummaryArea.get(summaryRoom);
+                            if (currentSummaryRoom.incomplete > 0) {
+                                //Goes through each record in the current room and updates the table with the info.
+                                ArrayList<String[]> myTableData = new ArrayList<>();
+                                ArrayList<String> myTableRowColors = new ArrayList<>();
+                                for (HashMap<String, String> currentRecord : currentSummaryRoom.incompleteRecords) {
+                                    //Updates the current row.
+                                    String[] myCurrentRow = new String[8];
+                                    myCurrentRow[0] = currentRecord.get("date");
+                                    myCurrentRow[1] = currentRecord.get("Id");
+                                    myCurrentRow[2] = currentRecord.get("tube_vol");
+                                    myCurrentRow[3] = currentRecord.get("tube_type");
+                                    myCurrentRow[4] = currentRecord.get("num_tubes");
+                                    myCurrentRow[5] = currentRecord.get("quantity");
+                                    myCurrentRow[6] = currentRecord.get("additionalServices");
+                                    myCurrentRow[7] = currentRecord.get("billedby/title");
+                                    myTableData.add(myCurrentRow);
+
+                                    //Colors the row yellow if there is 15 mins or less until record time, or red if the record time has passed.
+                                    String currentRowColor = "white";
+                                    LocalDateTime currentTime = LocalDateTime.now();
+                                    LocalDateTime timeLimit = currentTime.plusMinutes(15);
+                                    //Parses the record time because many of the records have optional milliseconds.
+                                    String recordTimeFormatted = currentRecord.get("date").split("\\.")[0];
+                                    LocalDateTime recordTime = LocalDateTime.parse(recordTimeFormatted, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
```

Comment:
So this one actually had me stumped for a while when I was coding it.

The DateToolkit is more for retrieving dates, whereas this is grabbing data from records.  For some reason, some of the records have milliseconds, and some do not.  Some record's milliseconds are also more digits than others.  I tried writing a function to convert them using a pattern, but because the date/time format changes for different records, my function kept throwing errors.  There was no catch-all method I could write to reformat the record's date/time without parsing out the milliseconds.